refactor: show all log entries

bafolts opened the pull request:
We have had an arbitrary limit placed on how much we log for game log entries. With #1110 there are users that want to see the entire log. I am curious to see how our memory will run on the heroku instance if we remove this limit. As long as this doesn't put us above the heroku memory limit it could be fine. I can keep an eye on the memory usage on heroku and remove if it is a problem.

chosta commented:
I'd suggest to create a function that gets the log by a given generation if possible.
Me and @ssimeonoff are working on an improved Log overview and he had the following idea:
Have an idicator of which generation is and load the log up to any other on clicking the corresponding number. I think this would solve the issue of memory overload as people would rarely want to see the whole log.
![log](https://user-images.githubusercontent.com/836179/94711421-c5755480-0350-11eb-86d9-077cbe4dda93.png)

kberg commented:
I'm also interested in storage, and latency impact.

On Wed, Sep 30, 2020, 12:13 PM Danimir <notifications@github.com> wrote:

> I'd suggest to create a function that gets the log by a given generation
> if possible.
> Me and @ssimeonoff <https://github.com/ssimeonoff> are working on an
> improved Log overview and he had the following idea:
> Have an idicator of which generation is and load the log up to any other
> on clicking the corresponding number. I think this would solve the issue of
> memory overload as people would rarely want to see the whole log.
> [image: log]
> <https://user-images.githubusercontent.com/836179/94711421-c5755480-0350-11eb-86d9-077cbe4dda93.png>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bafolts/terraforming-mars/pull/1428#issuecomment-701492715>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADE6KI6CKOWXSEFBUACL5TSINKMBANCNFSM4R7IN66A>
> .
>

bafolts commented:
OK. I will re-visit this. Having a separate api route to get the logs should avoid the latency impact. With some form of `/api/game/logs` that provides the logs we can avoid having to send every log to the client.

kberg commented:
:thumbsup

On Wed, Sep 30, 2020, 2:16 PM Brian Folts <notifications@github.com> wrote:

> OK. I will re-visit this. Having a separate api route to get the logs
> should avoid the latency impact. With some form of /api/game/logs that
> provides the logs we can avoid having to send every log to the client.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bafolts/terraforming-mars/pull/1428#issuecomment-701558750>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADE6KP7RSTXKEMU5WYUZ5LSINYZDANCNFSM4R7IN66A>
> .
>

bafolts commented:
@chosta I added a separate route for pulling logs of `/api/game/logs`. This change migrates the `LogPanel` to use the route. This should help with pulling logs by generation. This should help with latency and the entire log will never be sent to the client unless the client were to query for the entire log.

I also addressed two warnings which vue was logging to the developer console of missing key.

alrusdi commented on src/routes/Route.ts:
```diff
@@ -0,0 +1,31 @@
+
+import * as http from "http";
+
+export class Route {
```
Isn't class name little surprising?
As far as i know "routing" in web-apps is traditionally a way to map the Path to the View function.
All functions in this suggested class are Views actually.

Anyway it is great improvement to start the server.ts refactoring. This file takes alot of responsibility. It smells.

bafolts replied:
I have always referred to parts of an api as route. So the current `/api/game` would be the `GameRoute`. This class is meant to act as a utility of sorts to provide common responses that routes will need. There could be a better name here.

kberg replied:
Seems like a fine name, at least for now. A one-sentence comment at the top?

alrusdi replied:
> I have always referred to parts of an api as route. 

I'm worrying about naming if someday we will add **express** as server side framework. Their term Routing means something else than here https://expressjs.com/en/api.html#router But i'm not insisting of course - just trying to discuss it.

bafolts replied:
Using express makes sense and would help with people contributing. Should probably start to migrate towards express. As a history lesson I started this project out of boredom while commuting. I had a goal of taking on no dependencies. I don't think it would be too painful to wedge in express and use it for this new route while the old stuff used native `http`. Then eventually everything can port over to express. People are most likely more familiar with the documented express versus the jalopy code in server.ts

kberg reviewed (approved):
Having been in a shockingly high-latency network for a month (sattelite to a main house with a wifi repeater to the guest house, with a private VPN on my end because I don't think the owner understands network security) I do worry about this latency from a network perspective. Games took up to two minutes to load. (I told you it was bad.)

How often should that be your priority? Not a ton, I'm sure, but hundreds or thousand of log messages can be a bear to manage in all parts of the app.

I don't object to this change, but it's something to keep an eye on is all.

bafolts commented:
The `main.js` is 2.6MB since latency has never been a consideration. There is a quick win which should help you. We build with webpack setup in development mode. Using production mode cuts the build down to 1.36MB. Every player is downloading that file at least once prior to playback. I can put up a change to build in production mode if an environment variable is specified. The development build we use now comes with vue.js debug logging client side.

If you are running locally using a value of `production` here should give you a quick win.

https://github.com/bafolts/terraforming-mars/blob/master/webpack.config.js#L4

bafolts commented:
Merged this with master. Will manually test more tomorrow prior to merging. Will do work to introduce `express` for the server in a separate PR.
